### PR TITLE
RFC 0132 Monero Merge Mining

### PR DESCRIPTION
**Pull request by sdbondi**
## Description

RFC to describe the current protocol for monero merge mining.
[Rendered](https://demo.hedgedoc.org/s/y1dCJecba)

Note: there are fairly minor but consensus breaking differences in the transaction proof 
and other structs that will be PRd on the tari script branch.  

## Motivation and Context

RFC describing merge mining protocol were missing. Part of investigations into using a minimal merkle proof.

## How Has This Been Tested?

N/A

## Checklist:

* [x] I'm merging against the `development` branch.
* [x] I have squashed my commits into a single commit.

### CODE REVIEW

File: RFC/src/RFC-0132_Merge_Mining_Monero.md [hunk 1]
```diff
@@ -0,0 +1,163 @@
...
+```
+
+## Merge Mining Proxy 
+
+The [Tari merge mining proxy] proxies the [Monero daemon RPC] interface. It behaves as a middleware that implements the merge 
+mining protocol detailed above. This allows existing Monero miners to merge mine with Tari without having to make changes 
+to mining software. 
+
+The proxy must be configured to connect to a `monerod` instance, a Tari base node, and a Tari console wallet. Most requests
+are forwarded "as is" to `monerod`, however some are intercepted and augments before being returned to the miner.
```
Comment (delta1) [inline RFC/src/RFC-0132_Merge_Mining_Monero.md]:
```suggestion
are forwarded "as is" to `monerod`, however some are intercepted and augmented before being returned to the miner.
```

**Review by delta1 — DISMISSED** [review]
LGTM

File: RFC/src/RFC-0132_Merge_Mining_Monero.md [hunk 2]
```diff
@@ -0,0 +1,163 @@
...
+    /// Coinbase tx from Monero
+    pub coinbase_tx: MoneroTransaction,
+}
+```
+_fig 2. Monero PoW data serialized in Tari blocks
+
+A verifier may now check that the `coinbase_tx` contains the merge mining hash \\( h_m \\), and validate the `coinbase_merkle_proof` against the `transaction_root`.
+The `coinbase_merkle_proof` contains the minimal proof required to construct the `transaction_root`.
+
+For example, a tree with a depth of 2 will have a proof size of 3 * 32 bytes.
```
Comment (sdbondi) [inline RFC/src/RFC-0132_Merge_Mining_Monero.md]: Need to update this to reflect the changes made in upcoming PR - 3 * 32 bytes + 2 u32s used for the depth and path_bitmap

File: RFC/src/RFC-0132_Merge_Mining_Monero.md [hunk 3]
```diff
@@ -0,0 +1,172 @@
...
+## Introduction
+
+Tari employs a hybrid mining strategy, accepting 2 mining algorithms whose difficulties are independent of each other as discussed in [RFC-0131_Mining.html](/RFC-0131_Mining.html).
+This RFC details a protocol to enable Tari to accept Monero proof of work, enabling participating miners a chance to produce a valid block for either or both chain without additional mining effort.
+
+The protocol must enable a Tari base node to make the following assertions:
+
+**REQ 1.** The achieved mining difficulty exceeds target difficulty as dictated by Tari consensus, 
+
+**REQ 2.** The Monero block was constructed before the Tari block. This is to prevent a miner from submitting blocks from the parent chain that satisfy the auxiliary chain's difficulty without doing new work.
```
Comment (stringhandler) [inline RFC/src/RFC-0132_Merge_Mining_Monero.md]: Perhaps also:
The Monero block was constructed after the previous Tari block

File: RFC/src/RFC-0132_Merge_Mining_Monero.md [hunk 4]
```diff
@@ -0,0 +1,172 @@
...
+    /// References to the transactions included in this block
+    pub tx_hashes: Vec<hash::Hash>,
+}
+```
+_fig 1. The Monero block struct_
+
+Next, modify the Monero block template by including the merge mining hash \\( h_m \\) in the extra fields of the coinbase transaction. Monero has a [merge mining subfield] 
+to accommodate this data. Importantly, the extra field data part of the coinbase transaction hash and therefore the `merkle_tree_root`, the `blockhashing_blob` must be 
+reconstructed. A rust port of [Monero's tree hash] algorithm is needed to achieve this. The coinbase hash MUST be the first element to be hashed when constructing the `merkle_tree_root`.
+This satisfies **REQ 2**, proving that the Tari block was constructed before the parent chain's block.
```
Comment (stringhandler) [inline RFC/src/RFC-0132_Merge_Mining_Monero.md]: Is the order correct here? Seems to contradict REQ 2

File: RFC/src/RFC-0132_Merge_Mining_Monero.md [hunk 5]
```diff
@@ -0,0 +1,172 @@
...
+Concretely, A miner must serialize `MoneroPowData` using Monero consensus encoding and add it to the `pow_data` field in the Tari header.
+
+```rust,ignore
+pub struct MoneroPowData {
+    /// Monero header fields
+    header: MoneroBlockHeader,
+    /// randomX vm key
+    randomx_key: FixedByteArray, // Fixed 64 bytes
+    /// transaction count
+    transaction_count: u16,
```
Comment (stringhandler) [inline RFC/src/RFC-0132_Merge_Mining_Monero.md:112]: Might be more compact to serialize this with VarInt's

File: RFC/src/RFC-0132_Merge_Mining_Monero.md [hunk 6]
```diff
@@ -0,0 +1,172 @@
...
+    /// Coinbase tx from Monero
+    coinbase_tx: MoneroTransaction,
+}
+```
+_fig 2. Monero PoW data serialized in Tari blocks
+
+```rust,ignore
+pub struct MerkleProof {
+   branch: Vec<Hash>,
+   depth: u16,
```
Comment (stringhandler) [inline RFC/src/RFC-0132_Merge_Mining_Monero.md:126]: These could also use VarInts

**Review by stringhandler — COMMENTED** [review]
Really nice

File: RFC/src/RFC-0132_Merge_Mining_Monero.md (hunk 5 above)
Comment (sdbondi) [reply RFC/src/RFC-0132_Merge_Mining_Monero.md:112]: Will add some more details on the serialization and mention the VarInt encoding for all ints

File: RFC/src/RFC-0132_Merge_Mining_Monero.md (hunk 4 above)
Comment (sdbondi) [reply RFC/src/RFC-0132_Merge_Mining_Monero.md]: Maybe this is better

```suggestion
This satisfies **REQ 2**, proving that the proof-of-work was performed for the Tari block.
```

File: RFC/src/RFC-0132_Merge_Mining_Monero.md (hunk 6 above)
Comment (sdbondi) [reply RFC/src/RFC-0132_Merge_Mining_Monero.md:126]: Added a Serialisation section

File: RFC/src/RFC-0132_Merge_Mining_Monero.md (hunk 5 above)
Comment (sdbondi) [reply RFC/src/RFC-0132_Merge_Mining_Monero.md:112]: Added a Serialisation section

File: RFC/src/RFC-0132_Merge_Mining_Monero.md (hunk 4 above)
Comment (sdbondi) [reply RFC/src/RFC-0132_Merge_Mining_Monero.md]: Done

File: RFC/src/RFC-0132_Merge_Mining_Monero.md (hunk 3 above)
Comment (sdbondi) [reply RFC/src/RFC-0132_Merge_Mining_Monero.md]: Done

**Review by stringhandler — APPROVED** [review]
Nice

**Pull request merged**